src/components/views/dialogs/ReportEventDialog.tsx

```diff
@@ -0,0 +1,415 @@
+                        "should have a string field `content.room_id`, got", event);
+                    continue;
+                }
+                if (!("user_id" in content) || typeof content["user_id"] != "string") {
+                    // The room is improperly configured.
+                    // Display this debug message for the sake of moderators.
+                    console.debug("Moderation error", "state event", stateEventType,
+                        "should have a string field `content.user_id`, got", event);
+                    continue;
+                }
+                this.moderatedByRoomId = content["room_id"];
+                this.moderatedByUserId = content["user_id"];
+            }
+        }
+
+        this.state = {
+            // A free-form text describing the abuse.
+            reason: "",
+            busy: false,
+            err: null,
+            // If specified, the nature of the abuse, as specified by MSC3215.
+            nature: null,
+        };
+    }
+
+    _onReasonChange = ({target: {value: reason}}) => {
+        this.setState({ reason });
+    };
+
+    // The user has clicked on a nature.
+    _onNatureChosen = e => {
+        this.setState({ nature: e.target.value });
+    };
+
+    _onCancel = () => {
+        this.props.onFinished(false);
+    };
+
+    _onSubmit = async () => {
+        let reason = this.state.reason || "";
+        reason = reason.trim();
+        if (this.moderatedByUserId && this.moderatedByRoomId) {
+            // This room supports moderation.
+            // We need a nature.
+            // If the nature is `NATURE_OTHER` or `NATURE_ADMIN`, we also need a `reason`.
+            if (!this.state.nature ||
+                    ((this.state.nature == NATURE_OTHER || this.state.nature == NATURE_ADMIN)
+                        && !reason)
+            ) {
+                this.setState({
+                    err: _t("Please fill why you're reporting."),
+                });
+                return;
+            }
+        } else {
+            // This room does not support moderation.
+            // We need a `reason`.
+            if (!reason) {
+                this.setState({
+                    err: _t("Please fill why you're reporting."),
+                });
+                return;
+            }
+        }
+
+        this.setState({
+            busy: true,
+            err: null,
+        });
+
+        try {
+            if (this.moderatedByRoomId && this.moderatedByUserId && this.state.nature != NATURE_ADMIN) {
+                // Report to moderators through to the dedicated bot,
+                // as configured in the room's state events.
+                const client = MatrixClientPeg.get();
+                const dmRoomId = await ensureDMExists(client, this.moderatedByUserId);
+                const ev = this.props.mxEvent;
+                await MatrixClientPeg.get().sendEvent(dmRoomId, ABUSE_EVENT_TYPE, {
```

Comment:
Looks like there's a couple of places below where you can re-use the `client` variable defined on line 200